docs/nsm_api.md

```diff
@@ -0,0 +1,263 @@
+
+**interface list** defines supported/desired by POD connectivity types.
+
+**Note:** Interface structure also defined as a protobuf, for details see Appendix:
+
+- Connection Reply
+
+NSM Client connection reply message  is returned to inform the client if its request to NetworkService is successful or not.
+
+```proto
+message ConnectionReply {
+   bool accepted = 1;
+   string admission_error = 2;
+   ConnectionParameters connection_parameters = 3;
+   common.Interface interface = 4;
+}
+```
+**Where:**
+
+**accepted** true will indicate that the connection is accepted, otherwise false
+
+**admission\_error** will provide details why connection was refused.
+
+**interface** will indicate the selected/negotiated interface type
+
+**connection\_parameters** will provide interface specific parameters which the client is expected to parse and use.
+
+
+### NSE to its local NSM
+
+NSE is the actual provider of a network service, to make aware NSM of the service and some specific service parameters, NSE uses EndpointAdvertiseRequest message. Depending on NSE application, it can advertise multiple Network Services in a single message. NSM confirms acceptance of the advertisement in EndPointAdvertiseResponse message.
+
+- Endpoint AdvertiseRequest message
+
+```proto
+message EndpointAdvertiseRequest {
+   repeated netmesh.NetworkServiceEndpoint network\_endpoint = 1;
+}
+
+message NetworkServiceEndpoint {
+   string network_service_name = 1;
+   string network_service_host = 2;
+   string nse_provider_name = 3;
+   string nse_provider_namesapce = 4;
+   string socket_location = 5;
+   repeated common.Interface interface = 6;
+}
+```
+**Where:**
+
+**network\_service\_name** defines a name of the network service NSE supports
+
+**network\_service\_host** defines a name of a host where NSE runs
+
+**nse\_provider\_name** specifies NSE name, NSM used it to differentiate between multiple NSEs providing the same service
+
+**nse\_provider\_namespace** specifies NSE&#39;s kubernetes object namespace, NSM used it to differentiate between multiple NSEs providing the same service
+
+**socket\_location** informs NSM about linux named socket it has to use to communicate with NSE for connection requests
+
+### Local NSM to remote NSM not yet implemented
+
+When NSM local to NSM client discovers that NSE providing requested Network Service is not local, **network_service_host** in NSE custom resource object does not match the local NSM name, local NSM attempts to proxy client's request to remote NSM. gRPC over well known TCP socket is used for NSM to NSM communication. This method supports as "in-cluster" mode when NSM pod's DNS named is used as "out-of-cluster" when routable IP of external NSM is used to establish TCP connection. In order to facilitate NSM discovery for "in-cluster" mode, each NSM creates a kubernetes **Service** object with matching to NSM daemonset name. For each **Service**, a corresponding DNS entry is automatically created
```

Comment:
This paragraph is dense, and hard to understand at first reading. It seems it is more about NSM client wanting to connect to remote NSE, right?

Then, two cases:

- NSE (and its NSM) are in the cluster, and we use the Service IP provided by NSM;
- They are out of the cluster, and the external NSM IP is used.

I'm not even sure it is worth mentioning remote NSM/NSE for now? it seems far from being implemented.

Reply:
Ok, I will try to make it better :) but to clarify, NSM client (NS consumer) and NSE (NS provider) ALWAYS talk to its local NSM via linux socket. (Exception is the case of eNSM). If it happens that NS (Network Service) is not local to the client, only then NSM1 talks via tcp/ip to NSM2.

Client req. NSX -- linux socket --  NSM1 --  tcp/ip -- NSMX -- linux socket -- NSE (provider of NSX)

You are correct about two cases. About remote NSE, it is not my call :)